File: backend/huongvantra_backend/Service/ProductService/ProductService.Application/Validation/BomUnitRules.cs
using System.Globalization;
using System.Text;

namespace ProductService.Application.Validation;

public static class BomUnitRules
{
    private static readonly HashSet<string> CountBasedUnits = new(StringComparer.OrdinalIgnoreCase)
    {
        "cai",
        "chiec",
        "hop",
        "tui",
        "tem",
        "nhan",
        "goi",
        "chai",
        "lo"
    };

    private static readonly HashSet<string> MeasureBasedUnits = new(StringComparer.OrdinalIgnoreCase)
    {
        "gram",
        "g",
        "kg",
        "ml",
        "lit",
        "l"
    };

    public static bool IsCountBasedUnit(string? unit) =>
        CountBasedUnits.Contains(NormalizeUnit(unit));

    public static bool IsMeasureBasedUnit(string? unit) =>
        MeasureBasedUnits.Contains(NormalizeUnit(unit));

    public static bool IsIntegerQuantity(decimal quantity) =>
        decimal.Truncate(quantity) == quantity;

    public static bool HasMaxDecimalPlaces(decimal quantity, int maxDecimalPlaces = 3) =>
        quantity == decimal.Round(quantity, maxDecimalPlaces, MidpointRounding.ToZero);

    public static string NormalizeUnit(string? unit)
    {
        if (string.IsNullOrWhiteSpace(unit)) return string.Empty;

        var normalized = unit.Trim().ToLowerInvariant().Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(normalized.Length);

        foreach (var character in normalized)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) == UnicodeCategory.NonSpacingMark)
                continue;

            if (char.IsLetterOrDigit(character))
                builder.Append(character);
        }

        return builder.ToString().Normalize(NormalizationForm.FormC);
    }
}

File: frontend/huongvantra-web-client/src/features/products/utils/bomUnitRules.js
const COUNT_BASED_UNITS = new Set(['cai', 'chiec', 'hop', 'tui', 'tem', 'nhan', 'goi', 'chai', 'lo'])
const MEASURE_BASED_UNITS = new Set(['gram', 'g', 'kg', 'ml', 'lit', 'l'])

export function normalizeBomUnit(unit) {
  return String(unit ?? '')
    .trim()
    .toLowerCase()
    .normalize('NFD')
    .replace(/\p{Diacritic}/gu, '')
    .replace(/[^a-z0-9]/g, '')
}

export function isCountBasedUnit(unit) {
  return COUNT_BASED_UNITS.has(normalizeBomUnit(unit))
}

export function isMeasureBasedUnit(unit) {
  return MEASURE_BASED_UNITS.has(normalizeBomUnit(unit))
}

export function hasMaxDecimalPlaces(value, maxDecimalPlaces = 3) {
  const text = String(value ?? '').trim().replace(',', '.')
  if (!text || /e/i.test(text)) return false
  const decimalPart = text.split('.')[1] ?? ''
  return decimalPart.length <= maxDecimalPlaces
}

export function getBomQuantityValidationMessage(quantity) {
  const text = String(quantity ?? '').trim()
  if (!/^[1-9]\d*$/.test(text)) {
    return 'Định mức phải là số nguyên dương.'
  }

  return ''
}
